src/host-config/index.ts

```diff
@@ -0,0 +1,132 @@
+import { promises as fs } from 'fs';
+import { isRight } from 'fp-ts/lib/Either';
+import Reporter from 'io-ts-reporters';
+
+import { disallowedProxyFields } from './proxy';
+import type { RedsocksConfig, ProxyConfig } from './types';
+import { HostConfiguration, LegacyHostConfiguration } from './types';
+import { readProxy, setProxy } from './proxy';
+import * as config from '../config';
+import * as applicationManager from '../compose/application-manager';
+import { pathOnRoot } from '../lib/host-utils';
+import log from '../lib/supervisor-console';
+import * as updateLock from '../lib/update-lock';
+
+export * from './proxy';
+export * from './types';
+
+const hostnamePath = pathOnRoot('/etc/hostname');
+
+export async function readHostname() {
+	const hostnameData = await fs.readFile(hostnamePath, 'utf-8');
+	return hostnameData.trim();
+}
+
+export async function setHostname(val: string) {
+	let hostname = val;
+	// If hostname is an empty string, return first 7 digits of device uuid
+	if (!val) {
+		const uuid = await config.get('uuid');
+		hostname = uuid?.slice(0, 7) as string;
+	}
+	// Changing the hostname on config.json will trigger
+	// the OS config-json service to restart the necessary services
+	// so the change gets reflected on containers
+	await config.set({ hostname });
+}
+
+export class HostConfig {
+	public static patchProxy(
+		currentConf: RedsocksConfig,
+		inputConf: Partial<{
+			redsocks: Partial<ProxyConfig>;
+		}>,
+	): RedsocksConfig {
+		const patchedConf: RedsocksConfig = {};
+
+		// If input is empty, redsocks config should be removed
+		if (!inputConf || Object.keys(inputConf).length === 0) {
+			return patchedConf;
+		}
+
+		if (inputConf.redsocks && Object.keys(inputConf.redsocks).length > 0) {
+			// This method assumes that currentConf is a full ProxyConfig
+			// for backwards compatibility, so patchedConf.redsocks should
+			// also be a full ProxyConfig.
+			const patched = {
+				...currentConf.redsocks,
+				...inputConf.redsocks,
+			} as Record<string, any>;
+			// Delete local_ip, local_port in case a patch attempt for these fields is made
+			for (const field of disallowedProxyFields) {
+				delete patched[field];
+			}
+			patchedConf.redsocks = patched as ProxyConfig;
+		}
+		return patchedConf;
+	}
+
+	public static is(
+		conf: unknown,
+	): conf is HostConfiguration | LegacyHostConfiguration {
```

Comment:
We can't convert the input to HostConfiguration in all cases, since users may provide an input that does not decode to HostConfiguration:
```
t.type({
  network: t.partial({
    proxy: t.intersection([
      t.type({
        type: t.union([
          t.literal('socks4'),
          t.literal('socks5'),
          t.literal('http-connect'),
          t.literal('http-relay'),
        ]),
      }),
      t.partial({
        login: t.string,
        password: t.string,
        noProxy: t.array(t.string)
      }),
    ]),
    hostname: t.string
  })
})
```
but may decode to a shape that we historically accept (LegacyHostConfiguration):
```
t.type({
	network: t.partial({
		proxy: t.record(t.string, t.any),
		hostname: t.string,
	}),
})
```

Reply:
I don't understand, the LegacyHostConfiguration accepts pretty much any input? What would happen if an input was unused or invalid in that case? It doesn't seem compatible to me that we have one branch that properly validates and one that accepts anything. We can discuss about it in our checkin

Reply:
We discussed and agreed to keep this for now for backwards compatibility, and remove it in the next version of the Supervisor API.